File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    vector<vector<int>>ans;
    void solve(int idx, vector<int>& cand, vector<int>temp, int target){
        if(idx == cand.size()){
            if(target == 0){
                ans.push_back(temp);
            }
            return;
        }
        if(cand[idx] <= target){
            temp.push_back(cand[idx]);
            solve(idx, cand, temp, target - cand[idx]);
            temp.pop_back();
        }
        solve(idx+1, cand, temp, target);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        //Recursion and Backtracking 
        //TC - (2^t * k) 
        vector<int>temp;
        solve(0, candidates, temp, target);
        return ans;
    }
};

File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        n = len(candidates)
        def solve(idx, ss, temp):
            if idx >= n:
                return 
            if ss > target:
                return 
            if ss == target:
                ans.append(temp)
                return 
            solve(idx, ss+candidates[idx], temp+[candidates[idx]])
            solve(idx+1,ss,temp)

        solve(0,0,[])
        return ans